### PR TITLE
TPWDCEE-5915 - card expiration and type

### PR DESCRIPTION
**Pull request by flavius**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

File: Gateway/Model/Notify.php [hunk 1]
```diff
@@ -362,19 +382,29 @@ private function setCanCaptureInvoice($transactionType)
...
+        $responseData += ['card.0.card-type' => ''];
+        $cardType = $responseData['card.0.card-type'];
+        if (!empty($cardType)) {
+            $paymentToken->setType($cardType);
+        }
+        $cardType = $this->mapCardType($cardType);
+
         $paymentToken->setTokenDetails(json_encode([
-            'type' => '',
+            'type' => $cardType,
```
Comment (tomazpu) [inline Gateway/Model/Notify.php]: I would not create an extra variable i would use the function straight away. But no need to change.

File: Gateway/Model/Notify.php [hunk 2]
```diff
@@ -362,19 +382,29 @@ private function setCanCaptureInvoice($transactionType)
...
-        $paymentToken = $this->paymentTokenFactory->create();
+        /** @var PaymentTokenFactoryInterface $paymentToken */
+        $paymentToken = $this->paymentTokenFactory->create(PaymentTokenFactoryInterface::TOKEN_TYPE_CREDIT_CARD);
         $paymentToken->setGatewayToken($response->getCardTokenId());
         $paymentToken->setIsActive(true);
-        $paymentToken->setExpiresAt(date('d-m-Y', strtotime(date('d-m-Y', time()) . " + 1 year")));
         $paymentToken->setIsVisible(true);
         $paymentToken->setCustomerId($customerId);
         $paymentToken->setPaymentMethodCode($payment->getMethod());
+        $paymentToken->setExpiresAt($expirationDate);
```
Comment (tomazpu) [inline Gateway/Model/Notify.php]: I feel like the method is doing to much as it is creating a token, parsing the data and setting it.

File: Gateway/Model/Notify.php [hunk 3]
```diff
@@ -362,19 +382,29 @@ private function setCanCaptureInvoice($transactionType)
...
+
+        $responseData = $response->getData();
+        $responseData += ['card.0.card-type' => ''];
+        $cardType = $responseData['card.0.card-type'];
```
Comment (tomazpu) [inline Gateway/Model/Notify.php]: why don't we use the payment sdk 4.0.1 new getters to get the card type?

File: Gateway/Model/Notify.php [hunk 4]
```diff
@@ -390,6 +420,67 @@ protected function saveCreditCardToken($response, $customerId, $payment)
...
 
+    /**
+     * @param array $responseData
+     * @return array
+     */
+    private function extractCreditCardExpirationInformation(array $responseData)
+    {
+        $expirationYear = '';
+        $expirationMonth = '';
+        if (isset($responseData['card.0.expiration-year']) && isset($responseData['card.0.expiration-month'])) {
```
Comment (tomazpu) [inline Gateway/Model/Notify.php]: same here we are not using the getters in payment sdk for it, why?

File: Gateway/Model/Notify.php [hunk 5]
```diff
@@ -390,6 +420,67 @@ protected function saveCreditCardToken($response, $customerId, $payment)
...
+
+    /**
+     * @param SuccessResponse $response
+     * @throws \Exception
+     * @return string
+     * @since 3.1.0
+     */
+    private function createExpirationDate($response)
+    {
+        $responseData = $response->getData();
```
Comment (tomazpu) [inline Gateway/Model/Notify.php]: Same here you can use the card entity from paymentSDK to get the data

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

File: Gateway/Model/Notify.php [hunk 6]
```diff
@@ -481,4 +563,38 @@ protected function removeTokenByGatewayToken($customerId, $gatewayToken)
...
+    }
+
+    /**
+     * @param $response
+     * @param $customerId
+     * @param $payment
+     * @param string $expirationDate
+     * @return PaymentTokenFactoryInterface
+     */
+    private function createPaymentToken($response, $customerId, $payment, string $expirationDate): PaymentTokenFactoryInterface
```
Comment (rinnhofer) [inline Gateway/Model/Notify.php]: Codacy Issue found: [Line exceeds 120 characters; contains 127 characters](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

File: Gateway/Model/Notify.php (hunk 6 above)
Comment (rinnhofer) [inline Gateway/Model/Notify.php]: Codacy Issue found: [Line exceeds 120 characters; contains 127 characters](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- Gateway/Model/Notify.php  -4
- Model/Ui/TokenUiComponentProvider.php  -6
         

Complexity increasing per file
==============================
- Model/Ui/TokenUiComponentProvider.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5025879&bid=16562773)

**Pull request merged**